crates/l1tx/src/filter/indexer.rs

```diff
@@ -23,14 +23,17 @@ pub trait OpsVisitor {
     // Do stuffs with DA.
     fn visit_da<'a>(&mut self, _d: impl Iterator<Item = &'a [u8]>) {}
 
-    fn collect(self) -> Vec<ProtocolOperation>;
+    // Collect data
+    fn collect(
+        self,
+    ) -> (
+        Vec<ProtocolOperation>,
+        Vec<DepositRequestInfo>,
+        Vec<DaEntry>,
+    );
 }
 
 pub trait BlockIndexer {
-    type Output;
-
-    fn collect(self) -> Self::Output;
-
     fn index_tx(&mut self, txidx: u32, tx: &Transaction, config: &TxFilterConfig);
 
     fn index_block(mut self, block: &Block, config: &TxFilterConfig) -> Self
```

Comment:
The `index_tx` fn takes `&mut self` as would be expected, but why does this consume self?

Reply:
Oops, it doesn't have to.

Reply:
Oh wait, this was done to be able to chain the `.collect()` method. Or do you think it should just return `Self::Output` and get rid of the `collect()` method since we would always be `collect`ing it?

Reply:
Chaining isn't useful here, that's not how visitors are typically used.

Reply:
I don't think this has much to do with how visitors work. This totally respects how the visiting and accumulation works under the hood. It's just a matter of choice for finalizing the visitor imo. I prefer

```rust
let data = dataOpsIndexer::new()
    .index_block()
    .finalize();
```
to
```rust
let mut indexer = OpsIndexer::new();
indexer.index_block(&block);
let data = indexer.finalize();
```
We don't intend to do anything else with the indexer after calling `index_block()` except using the finalized data. So I think this chaining is fine.

Reply:
You might prefer it but it's a lot harder to understand the *purpose* of this chaining in context.

It *would* make sense if, by consuming the indexer, it returned the digested output, which then needed some finalization operation applied to it.  This is what's happening with iterator combinators, which wrap inner type with the combinator implementation type that exposes an implementation of `Iterator` based on the inner iterator type which is the next step up the chain.  But that's not what's happening here.